File: frontend/src/lib/chatApi.ts
// frontend/src/lib/chatApi.ts
import axios from 'axios';
import apiClient from '@/lib/api/clients'; // Import the shared axios instance

interface ChatRequest {
    message: string;
    conversation_id?: string;
}

export interface Message {
    createdAt: number;
    id: string;
    sender: 'user' | 'ai';
    content: string;
    timestamp: string;
    tool_calls?: any;
    tool_output?: any;
}

interface ChatResponse {
    conversation_id: string;
    message: string;
    ai_message: string;
    tool_invoked: boolean;
}

// Function to send a chat message
export const sendMessage = async (userId: string, chatRequest: ChatRequest): Promise<ChatResponse> => {
    try {
        const response = await apiClient.post<ChatResponse>(
            `/api/${userId}/chat`,
            chatRequest
        );
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            throw new Error(error.response.data.detail || 'Failed to send message.');
        }
        throw new Error('An unexpected error occurred while sending message.');
    }
};

// Function to fetch conversation history
export const fetchConversationHistory = async (userId: string): Promise<Message[]> => {
    try {
        const response = await apiClient.get<Message[]>(
            `/api/${userId}/chat/history`
        );
        return response.data;
    } catch (error) {
        if (axios.isAxiosError(error) && error.response) {
            throw new Error(error.response.data.detail || 'Failed to fetch conversation history.');
        }
        throw new Error('An unexpected error occurred while fetching conversation history.');
    }
};

File: backend/src/api/chat_api.py
from typing import Optional, List, Dict
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from src.models.user import User
from src.models.conversation import Conversation
from src.models.message import Message, Sender # Import Sender
from src.services.chat_service import ChatService
from src.services.ai_agent_service import AIAgentService # Import AIAgentService
from src.api.dependencies import get_current_user, verify_user_id, get_chat_service, get_ai_agent_service # Import new dependencies
from sqlmodel.ext.asyncio.session import AsyncSession
from src.models.database import get_async_session

router = APIRouter()

class ChatRequest(BaseModel):
    message: str
    conversation_id: Optional[UUID] = None

class ChatResponse(BaseModel):
    conversation_id: UUID
    message: str
    ai_message: str
    tool_invoked: bool = False # New field for T041

@router.post("/chat", response_model=ChatResponse)
async def chat_endpoint(
    request: ChatRequest,
    current_user: User = Depends(verify_user_id),
    chat_service: ChatService = Depends(get_chat_service),
    ai_agent_service: AIAgentService = Depends(get_ai_agent_service)
):
    """
    Handles chat messages from the user and returns an AI response.
    """
    conversation: Optional[Conversation] = None

    if request.conversation_id:
        conversation = await chat_service.get_conversation(request.conversation_id)
        if not conversation or conversation.user_id != current_user.id:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Conversation not found or not authorized."
            )
    else:
        conversation = await chat_service.create_conversation(current_user.id)

    # Add user message to conversation
    await chat_service.add_message(conversation.id, Sender.USER, request.message)

    # Get conversation history for AI agent
    messages = await chat_service.get_messages(conversation.id)

    # Generate AI response
    ai_response_content, tool_calls, tool_output = await ai_agent_service.generate_response( # Unpack new return type
        user=current_user,
        conversation=conversation,
        messages=messages,
        user_message=request.message
    )

    # Add AI response to conversation, including tool calls and outputs if any
    await chat_service.add_message(
        conversation.id,
        Sender.AI_AGENT,
        ai_response_content,
        tool_calls=tool_calls, # Pass tool_calls
        tool_output=tool_output # Pass tool_output
    )

    return ChatResponse(
        conversation_id=conversation.id,
        message=request.message,
        ai_message=ai_response_content,
        tool_invoked=bool(tool_calls) # Set based on whether tool_calls were made
    )